### PR TITLE
Check the stream position before checking if the cache is empty.

### PR DESCRIPTION
**Pull request by clokep**
If the stream position is at/earlier than the earliest known stream position, but the cache is empty we should still consider entities to have changed.

### CODE REVIEW

File: synapse/util/caches/stream_change_cache.py [hunk 1]
```diff
@@ -213,16 +213,17 @@ def has_any_entity_changed(self, stream_pos: int) -> bool:
         """
         assert isinstance(stream_pos, int)
 
-        if not self._cache:
```
Comment (clokep) [inline synapse/util/caches/stream_change_cache.py:216]: Note that `has_any_entity_changed` only has a single caller: `get_partial_current_state_deltas`, which is called from `get_current_state_deltas` which affects presence, stats, and the user directory.

Comment (clokep) [reply synapse/util/caches/stream_change_cache.py:216]: Searching the matrix.org database to see if there are descrepencies that could be due to this (or #14435 or #14592) shows that there are rooms where the room stats do not match current state events, this makes me think the user directory table might also be out of date? (I'll file a separate issue about this, I think.)

```sql
SELECT room_id, event_id, name AS stats_name, json::json->'content'->>'name' AS json_name
FROM current_state_events
INNER JOIN event_json USING (event_id, room_id)
INNER JOIN room_stats_state USING (room_id)
WHERE
    type = 'm.room.name' AND state_key = ''
    AND name != json::json->'content'->>'name';
```

Comment (clokep) [reply synapse/util/caches/stream_change_cache.py:216]: See #14643.

File: synapse/util/caches/stream_change_cache.py [hunk 2]
```diff
@@ -213,16 +213,17 @@ def has_any_entity_changed(self, stream_pos: int) -> bool:
...
-
         # _cache is not valid at or before the earliest known stream position, so
         # return that an entity has changed.
         if stream_pos <= self._earliest_known_stream_pos:
             self.metrics.inc_misses()
             return True
 
+        # If the cache is empty, nothing can have changed.
+        if not self._cache:
+            # TODO Metrics?
```
Comment (erikjohnston) [inline synapse/util/caches/stream_change_cache.py:224]: Yeah I think this should be a miss?

**Review by erikjohnston — APPROVED** [review]

File: synapse/util/caches/stream_change_cache.py (hunk 2 above)
Comment (clokep) [reply synapse/util/caches/stream_change_cache.py:224]:
```suggestion
            self.metrics.inc_misses()
```

File: changelog.d/14639.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix a long-standing bug where a device list update might not be sent to clients in certain circumstances.
```
Comment (clokep) [inline changelog.d/14639.bugfix:1]:
```suggestion
Fix a long-standing bug where the user directory and room/user stats might be out of sync.
```

File: synapse/util/caches/stream_change_cache.py [hunk 4]
```diff
@@ -213,16 +213,17 @@ def has_any_entity_changed(self, stream_pos: int) -> bool:
         """
         assert isinstance(stream_pos, int)
 
-        if not self._cache:
-            # If the cache is empty, nothing can have changed.
-            return False
-
         # _cache is not valid at or before the earliest known stream position, so
         # return that an entity has changed.
```
Comment (MadLittleMods) [inline synapse/util/caches/stream_change_cache.py:217]: We should add a comment here that the ordering is important. Something like:

> It's important that we check stream position for changes first before we do the early-return check against the cache below. If the stream position is at/earlier than the earliest known stream position, but the cache is empty we should still consider entities to have changed.

Comment (MadLittleMods) [reply synapse/util/caches/stream_change_cache.py:217]: @clokep Thoughts? Picking this up?

Comment (clokep) [reply synapse/util/caches/stream_change_cache.py:217]: I don't think this is necessary -- the updated comments are clear and the tests assert this.

Comment (MadLittleMods) [reply synapse/util/caches/stream_change_cache.py:217]: @clokep I don't pick any of this information up from the comments. Only from your PR description.

It's great that we have tests for this though.

**Pull request merged**